power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,240 @@ namespace detail = power_grid_model::optimizer::detail;
 
 using TrafoGraphIdx = Idx;
 using EdgeWeight = int64_t;
-using WeightedTrafo = std::pair<Idx2D, EdgeWeight>;
-using WeightedTrafoList = std::vector<WeightedTrafo>;
-using RankedTransformerGroups = std::vector<std::vector<Idx2D>>;
 constexpr auto infty = std::numeric_limits<Idx>::max();
+constexpr Idx2D unregulated_idx = {-1, -1};
+
+using enum Branch3Side;
+
+constexpr std::array<std::tuple<Branch3Side, Branch3Side>, 3> const branch3_combinations{
+    {{side_1, side_2}, {side_2, side_3}, {side_3, side_1}}};
```

Comment:
it's quite dangerous to do this `using enum` declaration here because the name for `side_1` etc. overlap between `Branch3Side` and `ControlSide`. In this case, I understand why you did it like this but it's probably better to scope it with a lambda. The main reason is that in the `tap_position_optimizer::detail` namespace, I would expect `ControlSide` to be the default used enum, not `Branch3Side`

```suggestion
using enum Branch3Side;

constexpr auto const branch3_combinations = []{
    using enum Branch3Side;
    return std::array<std::tuple<Branch3Side, Branch3Side>, 3>{
        {side_1, side_2}, {side_2, side_3}, {side_3, side_1}}};
```

in addition: is there a reason why you moved this outside the location where it is used? is it indeed used multiple times? if not, better to just move back to where it is used